### PR TITLE
IDEMPIERE-5567 Support of UUID as Foreign Key (FHCA-4195)

### PR DESCRIPTION
**Pull request by CarlosRuiz-globalqss**
# Pull Request Checklist

- [x] My code follows the [code guidelines](https://wiki.idempiere.org/en/Contributing_to_Trunk) of this project
- [x] My code follows the best practices of this project
- [x] I have performed a self-review of my own code
- [x] My code is easy to understand and review. 
- [x] I have checked my code and corrected any misspellings
- [x] In hard-to-understand areas, I have commented my code.
- [x] My changes generate no new warnings
### Tests
- [x] I have tested the direct scenario that my code is solving
- [x] I checked all collaterals that can be affected by my changes, and tested other potential affected scenarios
- [ ] New and existing unit tests pass locally with my changes
- [ ] I have added unit tests that prove my fix is effective or that my feature works
### Documentation
- [x] I have made corresponding changes to the documentation as follows:
- - [x] New feature (non-breaking change which adds functionality): I have created the New Feature page in the project wiki explaining the functionality and how to use it. If relevant, I have committed sample data to the core seed to have usable examples in GardenWorld.
- - [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected): I have documented the change in a clear way that everyone in the community can understand the impact of the change.
- - [ ] Improvement (improves and existing functionality): This documentation is needed if the improvement changes the way the user interacts with the system or the outcome of a process/task changes. If it is just, for instance, a performance improvement, documentation might not be needed. 
- [x] The changed/added documentation is in the project wiki (not privately-hosted pdf files or links pointing to a company website) and is complete and self-explanatory.

### CODE REVIEW

**Comment by hengsin** [comment]
It might be a good idea to take a look at the possibility to migrate _UU column to native UID column in PostgreSQL and Raw(16) in Oracle before adding foreign key support. Migration will be a lot more difficult once we started using it as foreign key as well.

**Comment by hengsin** [comment]
For PostgreSQL, to convert existing column from varchar to uuid type:
1. alter table test alter column test_uu drop default;
2. alter table test alter column test_uu type uuid using test_uu::uuid;

**Comment by hengsin** [comment]
http://crafted-software.blogspot.com/2013/03/uuid-values-from-jdbc-to-postgres.html (looks like getString wouldn't work).

**Comment by hengsin** [comment]
For Oracle:
ALTER TABLE test ADD test_uu_raw RAW(16);
UPDATE test SET test_uu_raw = hextoraw(replace(test_uu, '-', ''));
ALTER TABLE test DROP COLUMN test_uu;
ALTER TABLE test RENAME COLUMN test_uu_raw TO test_uu;

**Comment by CarlosRuiz-globalqss** [comment]
> It might be a good idea to take a look at the possibility to migrate _UU column to native UID column in PostgreSQL and Raw(16) in Oracle before adding foreign key support. Migration will be a lot more difficult once we started using it as foreign key as well.

Hi @hengsin - I'm afraid that will take too much time and is out of scope of my current deliverable.

I don't think a future migration will be harder because we already as a String.

By the way, because is not a SQL standard and lack of compatibility between different database, I think we must consider to keep managing the object as String on java side, and just change the way it is written/read on the database side.  But again, that's something that I wouldn't want to tackle at this moment.

**Comment by hengsin** [comment]
Migration is harder when foreign key constrain is in the picture.

One of the prime reason to use UUID key table is to store huge amount of data that's not possible with integer key. For that, the space and performance difference between storing UUID as native UID type (16 bytes) vs string (36 bytes) is pretty significant.

**Comment by CarlosRuiz-globalqss** [comment]
> Migration is harder when you add foreign key constrain to the picture.
> 
> One of the prime reason to use UUID key table is to store huge amount of data that's not possible with integer key. For that, the space and performance difference between storing UUID as native UID type (16 bytes) vs string (36 bytes) is pretty significant.

I mean, migrating from varchar to uuid (postgresql) sounds kind of "easy" as postgresql supports the alter within transaction.

Migrating that in oracle cab be a bit more sql involved as it implies dropping the constraints first, then migrating the data and then recreating back the constraint, but I don't see that as harder, just as extra sql steps.

I understand the advantages of native data types vs varchar (storage and index access mostly), but I also see there is a big problem about being non-standard SQL and big differences about how oracle and postgresql manage it.
Is doable, but my main problem at this moment is about the time that will be required to implement and test that.

**Review by hengsin — CHANGES_REQUESTED** [review]

File: org.adempiere.base/src/org/compiere/model/MZoomCondition.java [hunk 1]
```diff
@@ -244,16 +252,33 @@ public static int findZoomWindowByWindowId(int AD_Window_ID, MQuery query, int w
...
+								sql.append(parentTab.getTableName()).append("_ID=").append(parentId);
+							}
+							parentId = DB.getSQLValue(null, sql.toString());
+							if (parentTab.getLinkColumnName().endsWith("_UU")) {
+								parentId = DB.getSQLValueString(null, sql.toString());
+							} else {
+								int tmpId = DB.getSQLValue(null, sql.toString());
+								if (tmpId > 0)
+									parentId = Integer.valueOf(tmpId);
+							}
```
Comment (hengsin) [inline org.adempiere.base/src/org/compiere/model/MZoomCondition.java:270]: duplicate call to DB.getSQLValue ?

Comment (CarlosRuiz-globalqss) [reply org.adempiere.base/src/org/compiere/model/MZoomCondition.java:271]: Thanks @hengsin - fixed all the concerns

**Review by hengsin — APPROVED** [review]
looks good

**Pull request merged**